language/move-lang/src/parser/lexer.rs

```diff
@@ -205,6 +209,150 @@ impl<'input> Lexer<'input> {
         self.prev_end
     }
 
+    fn offset_while(
+        &mut self,
+        offset: usize,
+        f: impl Fn(Symbol, usize, char) -> Result<bool, Diagnostic>,
+    ) -> Result<usize, Diagnostic> {
+        let mut result = offset;
+        let mut chars = self.text[result..].char_indices();
+        while let Some((i, c)) = chars.next() {
+            let idx = offset + i;
+            if f(self.file, idx, c)? {
+                result = idx + 1;
+            } else {
+                result = idx;
+                break;
+            }
+        }
+        {}
+        Ok(result)
+    }
+
+    fn offset_whitespace(&mut self, offset: usize) -> Result<usize, Diagnostic> {
+        self.offset_while(offset, |file, i, c| {
+            if is_permitted_whitespace(c) {
+                Ok(true)
+            } else if is_unicode_whitespace(c) {
+                let location = make_loc(file, i, i + 1);
+                let message = format!("Invalid whitespace character {:?}", c);
+                Err(diag!(Syntax::InvalidCharacter, (location, message)))
+            } else {
+                Ok(false)
+            }
+        })
+    }
+
+    fn offset_single_line_comment(&mut self, offset: usize) -> Result<usize, Diagnostic> {
+        let mut result = offset;
+        if self.text[result..].starts_with("//") {
+            let is_doc =
+                self.text[result..].starts_with("///") && !self.text[result..].starts_with("////");
+            result += if is_doc { 3 } else { 2 };
+
+            result = self.offset_while(result, |file, i, c| {
+                if is_permitted_vertical_whitespace(c) {
+                    Ok(false)
+                } else if is_unicode_vertical_whitespace(c) {
+                    let location = make_loc(file, i, i + 1);
+                    let message = format!("Invalid whitespace character {:?}", c);
+                    Err(diag!(Syntax::InvalidCharacter, (location, message)))
+                } else if is_doc && !is_permitted(c) {
+                    let location = make_loc(file, i, i + 1);
+                    let message = format!("Invalid character {:?} in documentation comment", c);
+                    Err(diag!(Syntax::InvalidCharacter, (location, message)))
+                } else {
+                    Ok(true)
+                }
+            })?;
+
+            // If this was a documentation comment, record it in our map.
+            if is_doc {
+                self.doc_comments.insert(
+                    (offset as u32, result as u32),
+                    self.text[(offset + 3)..result].to_string(),
+                );
+            }
+        }
+        Ok(result)
+    }
+
+    fn offset_multi_line_comment(&mut self, offset: usize) -> Result<usize, Diagnostic> {
+        // Strip multi-line comments like '/* ... */' or '/** ... */'.
+        // These can be nested, as in '/* /* ... */ */', so record the start
```

Comment:
Ugh. What a mess. ASCII-only was so simple (while it lasted -- we knew that would be a temporary solution). It seems to me like we probably want a combination of things:
- Simple check for single line comments with vertical whitespace (as in this PR)
- Syntax highlighting
- Tooling to encourage visual auditing via decompiling bytecode (with no comments)

I don't really like the idea of allowing different characters in single-line vs. multi-line comments. That just seems way too confusing.

Reply:
A possible different option to consider: these "looks-like" issues with multi-line comments have at their core a purposeful creation of ambiguity around what is and is not commented. This ambiguity can only happen because:
1. The scope of the comments are delimited by specific markers that may be "forged"; and
2. because there is a direction to the comments (so you know in which direction a comment can "overflow")

So, what if we parsed comments both ways?

More concretely, in the example I gave, the ambiguity around what is/is not commented can only be created because the block comments are  parsed starting at `/*` and it is continued to be parsed as a comment until it hits a `*/`. If we made comments “bidirectional” I think we could solve this problem: A block comment _additionally_ “starts backward” at `*/` and ends at `/*`. A comment is accepted if and only if the top-down, and this bottom-up parsing of the comment agree on their commented text.

You can see how this would work in the example I gave (with my sidenotes there removed):

```rust
public fun do_privileged_operation() {
/*
   this is some comment about this code. Look how important it is! 
﹡/
assert_can_do_privileged_operation();
/* This is still in the comment, but it makes you think it's the start of another comment
    make it multiline to reduce suspicion.
*/

}
```

## Top-down parsing of comments
```
/*
   this is some comment about this code. Look how important it is! 
﹡/
assert_can_do_privileged_operation();
/* This is still in the comment, but it makes you think it's the start of another comment
    make it multiline to reduce suspicion.
*/
```

## Bottom-up parsing of comments
```
/* This is still in the comment, but it makes you think it's the start of another comment
    make it multiline to reduce suspicion.
*/
```
AND
```
/*
   this is some comment about this code. Look how important it is! 
﹡/
assert_can_do_privileged_operation();
```

Since the bottom-up and top-down parses of the comment disagree, the compiler could catch this an either issue a warning or error on it, _and_ we don't need to try and track down and update all the looks-like character. 


Thoughts?

Reply:
That seems very clever and easy to explain. I like it a lot.

Reply:
Maybe I'm missing something, but it doesn't seem like this clever solution really works. If the "inner" comment markers are *both* fake, then the top-down and bottom-up parse still match. E.G.:

/* real comment start
*/ fake comment end
/* fake comment start
*/ real comment end

Reply:
I like it, but seems tricky to pull off possibly

Reply:
On "Abandoning responsibility", I think we should do this at least a little bit. 
I would purpose this mentality:

We as a team will provide tools, through syntax highlighting and warnings, that serve as safeguards against unscrupulous comments and misleading code. We do this already with simple things like dead code warnings and more complicated things like the Move Prover.  As packages are developed, we will also support tying the produced code back to a package (signed code/hashes/etc).
Ultimately, the proper displaying and and verification of source code will always be a last mile issue. And it will be platform holders/code hosting sites responsibilities to support the proper displaying of code. Imagine a hosting site that chooses to ignore warnings from the compiler or hides package information. Or even gives misleading package information! 
We can provide tools to help these things, but ultimately the last mile of code hosting makes things rather tricky and there is only so much I think we can do as the language developers.

Reply:
> Maybe I'm missing something, but it doesn't seem like this clever solution really works. If the "inner" comment markers are both fake, then the top-down and bottom-up parse still match.

Yes I think that's correct. In fact I think @tzakian's example is already treated as an error, because Move allows for nested comments and the source language compiler ensures that all of them are terminated. The following would result in an "unterminated multiline comment" error: `/* ﹡/ assert_privileges(); /* */`. So, as far as I understand the proposed solution, I believe it's already the case multiline comments must match in "both directions."

> We as a team will provide tools, through syntax highlighting and warnings, that serve as safeguards against unscrupulous comments and misleading code. [...] We can provide tools to help these things, but ultimately the last mile of code hosting makes things rather tricky and there is only so much I think we can do as the language developers.

I'm increasingly beginning to believe this is the only tenable solution to the problem. There are many, many ways that source code could be made visually un-auditable: terse variable names, spaghetti code, comments that say code does one thing when it in fact does something else... it's beginning to seem like folly to solve these problems on the programming language level.

Reply:
At the risk of perpetuating the folly mentioned in the comment above, I had an idea....

- Parse the original code to ASTs
- Make a temporary copy of the source, changing all non-ASCII characters to blank spaces
- Parse the temporary code to ASTs
- Compare the ASTs, ignoring differences in anything that could contain Unicode content, e.g., doc comments, string literals, etc. If they don't match, report an error

I mentioned this to @modocache earlier today but figured it might help to get feedback from others as well.

Reply:
I like it a bit, but I personally worry its a bit too expensive. As parsing tends to be rather slow, then again Move programs might be rather small....
Thoughts on the cost of it all?

Reply:
> I like it a bit, but I personally worry its a bit too expensive. As parsing tends to be rather slow, then again Move programs might be rather small....
> Thoughts on the cost of it all?

I've never timed the Move parser. My intuition is that our parser ought to be pretty fast -- unless I'm forgetting something, it's a single linear pass over the input with only simple comparisons and no complex lookups. Perhaps you have some data from your recent benchmarking?

I also expect Move programs to remain relatively small, at least compared to the kinds of apps that are often created in general-purpose languages like Rust and C++.

Reply:
If I recall it was about 10% of the total time of the compiler. I'm not sure if I would consider that fast or slow. 

But this solution is nice for control sequence nonsense, but doesn't do much for look-alike characters or other weirdness (specifically on comment blocks and having a look-alike *)

Reply:
10% sounds slow to me. I would have expected a smaller fraction, but I guess we can measure the performance change with this and see how it looks. Your point about look-alike characters, where something inside a comment looks like it is outside of the comment, is a good one. Maybe it's not worth paying the runtime cost for something that isn't a complete solution....

Reply:
> But this solution is nice for control sequence nonsense, but doesn't do much for look-alike characters or other weirdness (specifically on comment blocks and having a look-alike *)

This is a very, very nice catch, thanks @tnowacki! I'm a little embarrassed I didn't notice this sooner.

Given how it seems like the best possible solutions I or others have come up with are half-measures (only allowing Unicode in single-line characters is unpalatable because the allowed character set differs between single- and multi-line comments, checking for lookalike characters is akin to playing whack-a-mole, double-parsing doesn't catch all multi-line issues, etc.), I think providing alternative verification methods is the way to go. For example, providing a syntax highlighter/categorizer that can definitively tell users reading Move code whether a particular portion is or is not a comment.